Add partial telemetry support

alexismanin opened the pull request:
Add code and docker-compose configuration for telemetry features:

 * Export logs with support for user geolocation (GeoIP)
 * Export traces

alexismanin committed:
fix(Docker): fix and document dev docker build

alexismanin committed:
feat(Compose): add a minimal compose configuration

alexismanin committed:
feat(Compose): Add observability stack

Add an observability configuration using Open Telemetry + Grafana to monitor application traces and logs

alexismanin committed:
feat(Python): activate uvicorn logging export using open-telemetry

alexismanin committed:
feat(Compose): Add an example dashboard for HTTP access statistics

alexismanin committed:
doc(Compose): Add documentation to help using and understanding compose configuration

Emile-Muller committed:
doc(compose): Changed code samples to use two backquotes instead of one (RST format).

alexismanin committed:
doc(Compose): fix image and link

J-Christophe commented on compose/README.rst:
Tu pourrais préciser la version de docker-compose que tu utilises.  Actuellement j'ai la *1.29.2*

`(planet_crs_registry-env) malapert@JC-HP-ZBook-15-G4:~/Work/planet_crs_registry/compose$ docker-compose --version
/usr/lib/python3/dist-packages/paramiko/transport.py:237: CryptographyDeprecationWarning: Blowfish has been deprecated
  "class": algorithms.Blowfish,
docker-compose version 1.29.2, build unknown`

Mais je suppose qu'elle trop vieille car j'ai *cette erreur* : 

`(planet_crs_registry-env) malapert@JC-HP-ZBook-15-G4:~/Work/planet_crs_registry/compose$ docker-compose -f compose.yaml -f observability.yaml up
/usr/lib/python3/dist-packages/paramiko/transport.py:237: CryptographyDeprecationWarning: Blowfish has been deprecated
  "class": algorithms.Blowfish,
ERROR: The Compose file './compose.yaml' is invalid because:
configs.api-conf value 'content' does not match any of the regexes: '^x-'
'name' does not match any of the regexes: '^x-'
You might be seeing this error because you're using the wrong Compose file version. Either specify a supported version (e.g "2.2" or "3.3") and place your service definitions under the `services` key, or omit the `version` key and place your service definitions at the root of the file to use version 1.
For more on the Compose file format versions, see https://docs.docker.com/compose/compose-file/`

Je viens d'installer une nouvelle version de docker-compose. Cela résoud le problème mais j'en ai un autre : 
$ /usr/local/bin/docker-compose version
          Docker Compose version v2.17.2
 $ /usr/local/bin/docker-compose up
          configs.api-conf Additional property content is not allowed

alexismanin replied:
J'ai testé avec les versions 2.23.3 et 2.27.1.
Je peux essayer de changer la façon dont les fichiers de configuration sont montés si tu veux, pour essayer d'être plus rétro-compatible avec d'anciennes versions.

alexismanin replied:
@J-Christophe : J'ai poussé un commit qui devrait améliorer la compatibilité du docker-compose avec une version un peu plus vieille de docker-compose. Dis-moi si tu as d'autres erreurs par commentaire.

J-Christophe replied:
Merci pour les changements. J'ai suivi le README mais il semble y avoir un bug losque je souhaite accéder à l'interface web http://localhost:8080/web :


`lanetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/starlette-0.27.0-py3.12.egg/starlette/middleware/exceptions.py", line 79, in __call__
planetary-crs-registry-api-1            |     raise exc
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/starlette-0.27.0-py3.12.egg/starlette/middleware/exceptions.py", line 68, in __call__
planetary-crs-registry-api-1            |     await self.app(scope, receive, sender)
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/fastapi-0.95.2-py3.12.egg/fastapi/middleware/asyncexitstack.py", line 21, in __call__
planetary-crs-registry-api-1            |     raise e
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/fastapi-0.95.2-py3.12.egg/fastapi/middleware/asyncexitstack.py", line 18, in __call__
planetary-crs-registry-api-1            |     await self.app(scope, receive, send)
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/starlette-0.27.0-py3.12.egg/starlette/routing.py", line 718, in __call__
planetary-crs-registry-api-1            |     await route.handle(scope, receive, send)
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/starlette-0.27.0-py3.12.egg/starlette/routing.py", line 276, in handle
planetary-crs-registry-api-1            |     await self.app(scope, receive, send)
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/starlette-0.27.0-py3.12.egg/starlette/routing.py", line 66, in app
planetary-crs-registry-api-1            |     response = await func(request)
planetary-crs-registry-api-1            |                ^^^^^^^^^^^^^^^^^^^
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/fastapi-0.95.2-py3.12.egg/fastapi/routing.py", line 237, in app
planetary-crs-registry-api-1            |     raw_response = await run_endpoint_function(
planetary-crs-registry-api-1            |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/fastapi-0.95.2-py3.12.egg/fastapi/routing.py", line 163, in run_endpoint_function
planetary-crs-registry-api-1            |     return await dependant.call(**values)
planetary-crs-registry-api-1            |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/planet_crs_registry-0.9.5.dev37+gec46c5d-py3.12.egg/planet_crs_registry/core/routers/ws_router.py", line 143, in get_versions
planetary-crs-registry-api-1            |     await WKT_model.all()
planetary-crs-registry-api-1            |           ^^^^^^^^^^^^^^^
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/tortoise_orm-0.19.3-py3.12.egg/tortoise/models.py", line 1281, in all
planetary-crs-registry-api-1            |     db = using_db or cls._choose_db()
planetary-crs-registry-api-1            |                      ^^^^^^^^^^^^^^^^
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/tortoise_orm-0.19.3-py3.12.egg/tortoise/models.py", line 1033, in _choose_db
planetary-crs-registry-api-1            |     db = router.db_for_read(cls)
planetary-crs-registry-api-1            |          ^^^^^^^^^^^^^^^^^^^^^^^
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/tortoise_orm-0.19.3-py3.12.egg/tortoise/router.py", line 36, in db_for_read
planetary-crs-registry-api-1            |     return self._db_route(model, "db_for_read")
planetary-crs-registry-api-1            |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/tortoise_orm-0.19.3-py3.12.egg/tortoise/router.py", line 31, in _db_route
planetary-crs-registry-api-1            |     return connections.get(self._router_func(model, action))
planetary-crs-registry-api-1            |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
planetary-crs-registry-api-1            |   File "/usr/local/lib/python3.12/dist-packages/tortoise_orm-0.19.3-py3.12.egg/tortoise/router.py", line 18, in _router_func
planetary-crs-registry-api-1            |     for r in self._routers:
planetary-crs-registry-api-1            | TypeError: 'NoneType' object is not iterable
planetary-crs-registry-api-1            | HTTP Request: GET http://localhost:8080/ws/versions "HTTP/1.1 500 Internal Server Error"

En revanche, cette URL fonctionne : http://localhost:8080/ws/IAU/2015/1000 mais celle-ci ne fonctionne plus http://localhost:8080/ws/IAU/2015 (internal server error)

@alexismanin, même comportement chez toi ?

Merci

alexismanin replied:
@J-Christophe : effectivement, je n'avais pas vu ce problème, mais je le reproduis aussi. Je vais regarder ce qui se passe. Je n'ai pas touché grand-chose dans le code, alors je ne comprends pas encore ou est la régression. Je vais chercher et je te tiens au courant.

J-Christophe replied:
@alexismanin j'ai trouvé, il y a un pb dans le fichier de conf avec le nom de la base GEOPI. La base  doit s'écrire de cette manière : GeoLite2-City.mmdb (je l'ai récupéré ainsi dans le zip). Dans le fichier de conf, tout est en minuscule d'où le problème. Tout démarre sans erreur maintenant. 

2 questions : 
- je ne vois aucun point sur la carte lorsque je teste. J'imagine car je teste en local 
- il y a moyen de supprimer dans "HTTP request count by path and method" toutes les URLs qui commencent par /assets/ et /favicon.ico (ce ne sont pas des URLs métiers)

alexismanin replied:
> @alexismanin j'ai trouvé, il y a un pb dans le fichier de conf avec le nom de la base GEOPI. La base doit s'écrire de cette manière : GeoLite2-City.mmdb (je l'ai récupéré ainsi dans le zip). Dans le fichier de conf, tout est en minuscule d'où le problème. Tout démarre sans erreur maintenant.

Pour la bdd GeoIP, le readme précise de la renommer pour que ça matche le fichier de conf (mais on peut faire l'inverse si on veut, le tout c'est de le savoir).
Cependant, cela n'impacte pas les API HTTP, car la bdd geoip est utilisée en aval par un service tiers (Grafana Alloy).
L'erreur sur les endpoints `/web` et `/ws/IAU/2015` vient d'ailleurs (je la reproduis même sans télémétrie activée).
Je regarde en ce moment, mais j'ai l'impression que le pb vient du fait que j'ai utilisé la fonctionnalité "lifespan" de FastAPI au lieu du mécanisme déprécié "on_event". Je pense que d'autres composants doivent utiliser le même mécanisme, et la doc spécifie qu'on ne peut pas utiliser les deux mécanismes en même temps.

> 
> 2 questions :
> 
>     * je ne vois aucun point sur la carte lorsque je teste. J'imagine car je teste en local

Exactement. C'est la partie un peu délicate. Pour tester ça, il faut déployer le service sur une IP publique, et taper dessus depuis l'éxtérieur.

>     * il y a moyen de supprimer dans "HTTP request count by path and method" toutes les URLs qui commencent par /assets/ et /favicon.ico (ce ne sont pas des URLs métiers)

Oui, il faut éditer le graphique concerné (menu en haut à droite sur le graphique concerné), et ensuite il faut ajouter une transformation de type "filter by field", ou qqch comme ça, pour filtrer sur le champs "uri_path" (et donc, omettre les chemins concernés).

alexismanin replied:
Je confirme, et c'est vicieux : le bug se produit à cause du "lifespan" que j'ai utilisé, mais seulement après un premier redémarrage du service (c'est peut-être pour ça que je n'avais pas vu le pb jusqu'ici). En fait, il y a un autre code dans le projet qui utilise le mécanisme de "on_event" pour fermer la connexion à la bdd quand on éteint le service. L'utilisation du "lifespan" empếche la bonne fermeture, et donc, j'imagine, corromps la bdd sqlite.
Bref, j'ai changé mon code pour utiliser "on_event" au lieu de "lifespan", ça corrige le pb.

J-Christophe replied:
Je ne m'en sors pas avec grafana pour supprimer l'ensemble des /asset. J'ai fait ceci mais rien ne se passe 

![Edit-panel-HTTP-accesses-Dashboards-Grafana](https://github.com/pdssp/planet_crs_registry/assets/2744725/88fd91d4-cce9-46ae-8cd8-827a7c9bf4f4)

Une idée ?

J-Christophe replied:
Trouvé. Je n'avais pas vu qu'il y avait d'autres options dans "Match". Tout est OK

alexismanin committed:
refactor(Compose): change api configuration binding

Use bind mount instead of config block, to enhance retro-compatibility with older versions of docker-compose

alexismanin committed:
fix(Python): refactor otel/uvicorn binding to use "on_event('startup')" feature instead of lifespan.

Use of lifespan broke tortoise shutdown (which use on_event("shutdown") in ws_router file.
